cli/azd/pkg/containerapps/container_app.go

```diff
@@ -116,107 +159,166 @@ func (cas *containerAppService) AddRevision(
 	if err != nil {
-		return fmt.Errorf("updating container app revision: %w", err)
+		return nil, fmt.Errorf("updating container app revision: %w", err)
 	}
 
-	// If the container app is in multiple revision mode, update the traffic to point to the new revision
-	if *containerApp.Properties.Configuration.ActiveRevisionsMode == armappcontainers.ActiveRevisionsModeMultiple {
-		newRevisionName := fmt.Sprintf("%s--%s", appName, *revision.Properties.Template.RevisionSuffix)
-		err = cas.setTrafficWeights(ctx, subscriptionId, resourceGroupName, appName, containerApp, newRevisionName)
-		if err != nil {
-			return fmt.Errorf("setting traffic weights: %w", err)
-		}
+	newRevisionName := fmt.Sprintf("%s--%s", appName, *revision.Properties.Template.RevisionSuffix)
+	newRevisionResponse, err := revisionsClient.GetRevision(ctx, resourceGroupName, appName, newRevisionName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting new revision '%s': %w", newRevisionName, err)
 	}
 
-	return nil
+	return &newRevisionResponse.Revision, nil
 }
 
-func (cas *containerAppService) syncSecrets(
+// Validates that a revision is ready to accept traffic.
+// This method will block until the revision is ready, reaches a terminal state or times out
+// Expects that revision to be in the following states:
+// ProvisioningState: "Provisioned"
+// HealthState: "Healthy"
+func (cas *containerAppService) ValidateRevision(
 	ctx context.Context,
 	subscriptionId string,
 	resourceGroupName string,
 	appName string,
-	containerApp *armappcontainers.ContainerApp,
-) (*armappcontainers.ContainerApp, error) {
-	// If the container app doesn't have any secrets, we don't need to do anything
-	if len(containerApp.Properties.Configuration.Secrets) == 0 {
-		return containerApp, nil
-	}
-
-	appClient, err := cas.createContainerAppsClient(ctx, subscriptionId)
+	revisionName string,
+) (*armappcontainers.Revision, error) {
+	revisionsClient, err := cas.createRevisionsClient(ctx, subscriptionId)
 	if err != nil {
 		return nil, err
 	}
 
-	// Copy the secret configuration from the current version
-	// Secret values are not returned by the API, so we need to get them separately
-	// to ensure the update call succeeds
-	secretsResponse, err := appClient.ListSecrets(ctx, resourceGroupName, appName, nil)
-	if err != nil {
-		return nil, fmt.Errorf("listing secrets: %w", err)
-	}
+	var result *armappcontainers.Revision
 
-	secrets := []*armappcontainers.Secret{}
-	for _, secret := range secretsResponse.SecretsCollection.Value {
-		secrets = append(secrets, &armappcontainers.Secret{
-			Name:  secret.Name,
-			Value: secret.Value,
-		})
-	}
+	// Wait 10 seconds between retries
+	backOffDuration := retry.NewConstant(time.Second * 10)
 
-	containerApp.Properties.Configuration.Secrets = secrets
+	// Keep trying for up to 10 minutes or has reached a terminal state
+	maxDuration := retry.WithMaxDuration(time.Minute*10, backOffDuration)
```

Comment:
Being a user of `az`, there are multiple things that frustrates users such as myself that I'd love to avoid in `azd` that isn't solved in this current implementation:

1. The polling timeout is set to higher-than-expected durations (10m), where the CLI just spins for 10 minutes when an unrecoverable error occurs. This is extremely frustrating as an end-user.
1. There is little interactivity that shows that progress is being made. Is the CLI stuck? How does a user know?
1. When an error occurs, there are no hints as to whether:
   1. A user error occurred
   1. A service error occurred
   1. The CLI has a bug in it

10 minutes is also a forever time scale for most applications. I expect AKS to perform a lot better, in which this is set too high. I would also love for this to be smarter in handling different kinds of error conditions.

I'd love for us to spec this out more, and more generally for all deployment hosts. To me, the UX is the crucial thing we should get right more than the feature itself. Curious what @ellismg and @savannahostrowski thinks here.

Reply:
All of the above points seem like big UX pain points. Thanks for listing them out, @weikanglim.

Is there opportunity for us to leverage existing patterns here to provide more of a signal around progress? Probably some eng work to provide that level of status. Also, would love to see actionable error messages here (and also elsewhere)!

Reply:
If an unrecoverable error occurs such as transitioning into  `ProvisionError`, the operation will fail faster than the 10 minutes window.

Health checks are are more tricky.  During startup I it could take some time and a health probe could transition from Unhealth to Healthy within a span of minutes depending on what the health probes validate internally.

I am open to constructive suggestions.

Reply:
It is interesting to consider application health checks. If the health check times out after 10 minutes, we fail the deployment operation, but the deployment is technically ongoing, which is a really odd thing. Personally, health checks should be something the user opts in to, because they know what warm-up SLAs they've set up for themselves, and also what do to in case of health check failures, which includes timeouts or direct failures. I suspect they want it configurable to rollback if necessary.

Having said all that, we should think about what all this means for a simple, normative use-case -- our python and JS developers. Do we want health checks by default? If so, what is the acceptable experience? And certainly, consider how we communicate errors back. For example, platform failures (provision) should be treated separately than application startup failures.